---
title: "Introduction to iGC"
author: "iGC Developers"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{Introduction to iGC}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

With the advancement in high-throughput technologies, researchers can investigate gene expression and copy number alteration (CNA) data of one patient at a lower cost. Traditional analysis method is to analyze each platform individually and integrate their results using Venn diagrams. Challenges arise, however, when the low reproducibility and inconsistency are observed across multiple platforms. To address these issues, we developed an integrated analysis package to concurrently analyze Gene expression profiles and Copy number changes (`iGC`) in the same samples. The results showed `iGC` can facilitate identification of CNA-driven genes and their Pearson correlation coefficients were significantly higher than those genes located in copy number alteration regions. Therefore, it is a better approach to simultaneously consider changes in both genome and transcriptome in an integrated analysis. All input and output formats for the iGC package are tab-delimitated files, and the input files can be directly imported from The Cancer Genome Atlas (TCGA), Gene Expression Omnibus (GEO), and other sources.

## Get gene expression list of entire input samples

The `create_gene_exp` function is used to rearrange the input gene expression files from The Cancer Genome Atlas (TCGA), Gene Expression Omnibus (GEO), and other sources to get an overall gene expression list of entire samples.

## Map CNA data to human genes using hg19-Refseg

After submitting, the CN data will be mapped to human genes and the reference is human genome version 19. The human genome reference used here is RefSeq transcripts in version hg19 from UCSC Genome Browser. The transcripts with NM marker ID, which are protein-codeing, were selected to be our reference database and provided as hg19DBNM.rda. This reference provides region information, including chromosome number, starting position, ending position, strand and gene symbols, for converting copy number alteration data into human genes. 
Next, the mapped human genes will be defined as CN gain or loss based on the CN threshold, whose default values are set as 2.5 for gain and 1.5 for loss. These mapped genes will be assigned values in +1, -1 or 0, where +1 stands for CNA-gain, -1 stands for CNA-loss and 0 stands for neutral.
Then the `faster_gene_cna` function will return a gene list of its CNA status, CNA-gain (G), CNA-loss (L) and neutral (N), in all input samples and the ratio of the number of CN changed samples (G or L) to the number of total samples.

## Identifies CNA-driven differentially expressed genes

After the input samples are divided into CNA-gain (G), CNA-loss (L), and neutral (N) groups based on its CNA statuses for each gene, one gene remains for further analyses if its ratio of the number of CN changed samples (G or L) to the number of total samples is larger than a given threshold. Here, defaultly only genes showing CNAs in at least 20% of the samples will be analyzed further by the `find_cna_drive_gene` function, but the threshold can be changed by users. Lastly, to take CNA into consideration, statistical tests, Studentâ€™s t-test and Wilcoxon rank sum test, are performed in the GE level by classifying the samples as G and L plus N groups or L and G plus N groups, depending on the CN of the interested gene increases or decreases. False discovery rate, p-value, associated statistics are included in output files, and gene set enrichment analysis (GSEA) can be performed directly.
