---
title: "Analysis Example using iGC"
author: "iGC Developers"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

In this section we will use `iGC` to perform an analysis on a subset of the breast cancer dataset selecting from The Cancer Genome Atlas (TCGA) in the format of level three. The example dataset contains 50 breast cancer samples with both its gene expression and copy number alterations.

## Install Package

To begin with, we install the required package through the biocLite(), and then load it.

```{r, eval=FALSE}
source("http://www.bioconductor.org/biocLite.R")
biocLite("iGC")
```

```{r}
library(iGC)
```

## Generate Sample Description

Next, before analyzing, a sample description table that contains IDs of gene expression and copy number files for each sample is needed for mapping. The create_sample_desc function is provided for creating sample description table containing all required inputs. 

```{r}
sample_desc_pth <- system.file("extdata", "sample_desc.csv", package = "iGC")
sample_desc <- create_sample_desc(sample_desc_pth)
head(sample_desc)
```

```{r, eval=FALSE}
sample_desc_alt <- create_sample_desc(
  sample_names = sample_desc$Sample,
  cna_filepaths = sample_desc$CNA_filepath,
  ge_filepaths = sample_desc$GE_filepath
)
head(sample_desc_alt)
```

## Get oveall GE list using function `create_gene_exp`

Subsequently, we apply the `create_gene_exp` function to create an overall gene expression list of entire input samples. The output file will be saved as gene_exp.rda.

```{r}
gene_exp <- create_gene_exp(sample_desc, progress = FALSE)
gene_exp[GENE %in% c('TP53', 'BRCA1', 'NFKB1'), 1:10, with=FALSE]
```

```{r, eval=FALSE}
gene_exp <- create_gene_exp(
  sample_desc,
  read_fun = read.table,
  progress = TRUE, progress_width = 60,
  header = FALSE,
  skip = 2,
  na.strings = "null",
  colClasses = c("character", "double")
)
```

## Map CNA to human genes using function `faster_gene_cna`

Then, we run `faster_gene_cna` function to map CNA data to human genes using human genome version 19 as reference. For each mapped human genes, the input samples are divided into CNA-gain (G), CNA-loss (L), and neutral (N) groups based on its CNA statuses, whose default values are set as 2.5 for gain and 1.5 for loss. The output file will be saved as gene_cna.rda.


```{r}
my_cna_reader <- function(cna_filepath) {
  cna <- data.table::fread(
    cna_filepath, sep = '\t', header = TRUE
  )
  cna[, .(Chromosome, Start, End, Segment_Mean)]
}

<<<<<<< HEAD
gain_loss = log2(c(2.5, 1.5)) - 1
gene_cna <- faster_gene_cna(
  as.data.table(sample_desc),
  cna.gain.threshold = gain_loss[1], cna.loss.threshold = gain_loss[2],
  progress = TRUE, progress_width = 60,
  read_fun = my_cna_reader
=======
gain_loss = log2(c(2.4, 1.6)) - 1
gene_cna <- create_gene_cna(
  sample_desc,
  gain_threshold = gain_loss[1], loss_threshold = gain_loss[2],
  read_fun = my_cna_reader,
  progress = FALSE
>>>>>>> bb0b69bc216f60ef9bb81b67b05100452d087645
)
gene_cna[GENE %in% c('TP53', 'BRCA1', 'NFKB1'), 1:10, with=FALSE]
```

```{r, eval=FALSE}
doMC::registerDoMC(cores = 4)
gene_cna <- faster_gene_cna(
  sample_desc, gain_loss[[1]], gain_loss[[2]], parallel = TRUE
)
```


## Identify CNA-driven differentially expressed genes using function `find_cna_driven_gene`

Lastly, we employ `find_cna_driven_gene` function to identify differentially expressed genes driven by CNAs from samples with both proprocessed GE and CNA data that were obtained by previous two steps. Genes of gene_cna remain for further analyses if its ratio of the number of CN changed samples (G or L) to the number of total samples is larger than a given threshold. Then, statistical tests will be performed in the GE level by classifying the samples as G and L plus N groups or L and G plus N groups, depending on the CN of the interested gene increases or decreases. That is, only genes showing CNAs in at least 20% of the samples will be analyzed further, and the threshold can be changed by users. The GE levels are evaluated by the Studentâ€™s t-test and Wilcoxon rank sum test. False discovery rate, p-value, associated statistics are included in output files, and gene set enrichment analysis (GSEA) can be performed directly.

```{r}
cna_driven_genes <- find_cna_driven_gene(
  gene_cna, gene_exp,
  gain_ratio = 0.15, loss_ratio = 0.15,
  progress = FALSE, parallel = FALSE
)
head(cna_driven_genes$gain_driven)
head(cna_driven_genes$loss_driven)
head(cna_driven_genes$both)
```

```{r}
cna_driven_genes$loss_driven[GENE %in% c('BRCA1')]
```
